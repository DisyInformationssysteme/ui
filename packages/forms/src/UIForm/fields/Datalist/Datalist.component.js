import React, { Component } from 'react';
import PropTypes from 'prop-types';
import DataListComponent from '@talend/react-components/lib/Datalist';
import omit from 'lodash/omit';
import FieldTemplate from '../FieldTemplate';

export function escapeRegexCharacters(str) {
	return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

const PROPS_TO_OMIT = [
	'classNames',
	'schema',
	'getComponent',
	'errorMessage',
	'isValid',
	'onChange',
	'onFinish',
	'onTrigger',
	't',
];

const SCHEMA_TO_OMIT = ['type', 'triggers', 'title', 'titleMap', 'schema'];

class Datalist extends Component {
	constructor(props) {
		super(props);
		this.state = {};
		this.onChange = this.onChange.bind(this);
		this.onFocus = this.onFocus.bind(this);
		this.getTitleMap = this.getTitleMap.bind(this);
		this.isMultiple = this.isMultiple.bind(this);
		this.callTrigger = this.callTrigger.bind(this);
	}

	componentDidMount() {
		this.callTrigger({ type: 'didMount' });
	}
	/**
	 * On change callback
	 * We call onFinish to trigger validation on datalist item selection
	 * @param event
	 * @param payload
	 */
	onChange(event, payload) {
		const payloadWithSchema = { ...payload, schema: this.props.schema };
		this.callTrigger(event);
		this.props.onChange(event, payloadWithSchema);
		this.props.onFinish(event, payloadWithSchema);
	}

	onFocus(event) {
		this.callTrigger(event);
	}

	onBlur(event) {
		this.callTrigger(event);
	}

	getTitleMap() {
		let titleMap;
		if (this.state.titleMap) {
			titleMap = this.state.titleMap;
		} else if (this.props.schema.titleMap) {
			titleMap = this.props.schema.titleMap;
		} else {
			titleMap = [];
		}

		const values = this.isMultiple() ? this.props.value : [this.props.value];
		const optionNotExists = value => titleMap.find(option => option.value === value);
		const additionalOptions = values.filter(optionNotExists).reduce((acc, value) => {
			acc.push({ name: value, value });
			return acc;
		}, []);
		// console.log({ additionalOptions });
		return titleMap.concat(additionalOptions);
	}

	isMultiple() {
		return this.props.schema.schema.type === 'array';
	}

	callTrigger(event) {
		if (this.props.schema.triggers) {
			const trigger = this.props.schema.triggers.find(t => t.onEvent === event.type);
			if (trigger) {
				this.setState({ isLoading: true });
				this.props.onTrigger(event, {
					trigger,
					schema: this.props.schema,
					errors: this.props.errors,
					properties: this.props.properties,
				}).then(data => {
					this.setState({
						isLoading: false,
						...data,
					});
				}, () => {
					this.setState({ isLoading: false });
				});
			}
		}
	}

	render() {
		const props = omit(this.props, PROPS_TO_OMIT);
		Object.assign(props, this.state);
		Object.assign(props, omit(this.props.schema, SCHEMA_TO_OMIT));
		props.titleMap = this.getTitleMap();
		if (props.id) {
			props.id = `${props.id}-select`;
		}
		return (
			<FieldTemplate
				description={this.props.schema.description}
				errorMessage={this.props.errorMessage}
				id={this.props.id}
				isValid={this.props.isValid}
				label={this.props.schema.title}
				required={this.props.schema.required}
			>
				<DataListComponent
					{...props}
					input
					multiSection={false}
					onChange={this.onChange}
					onFocus={this.onFocus}
				/>
			</FieldTemplate>
		);
	}
}

Datalist.displayName = 'Datalist field';
Datalist.defaultProps = {
	value: '',
};

if (process.env.NODE_ENV !== 'production') {
	Datalist.propTypes = {
		id: PropTypes.string,
		isValid: PropTypes.bool,
		errorMessage: PropTypes.string,
		onChange: PropTypes.func.isRequired,
		onFinish: PropTypes.func.isRequired,
		onTrigger: PropTypes.func,
		errors: PropTypes.object,
		properties: PropTypes.object,
		schema: PropTypes.shape({
			schema: PropTypes.shape({
				type: PropTypes.string,
			}),
			triggers: PropTypes.arrayOf(
				PropTypes.shape({
					onEvent: PropTypes.string,
				}),
			),
			autoFocus: PropTypes.bool,
			description: PropTypes.string,
			disabled: PropTypes.bool,
			placeholder: PropTypes.string,
			readOnly: PropTypes.bool,
			required: PropTypes.bool,
			restricted: PropTypes.bool,
			title: PropTypes.string,
			titleMap: PropTypes.arrayOf(
				PropTypes.shape({
					name: PropTypes.string.isRequired,
					value: PropTypes.string.isRequired,
				}),
			),
		}),
		value: PropTypes.string,
	};
}

export default Datalist;
