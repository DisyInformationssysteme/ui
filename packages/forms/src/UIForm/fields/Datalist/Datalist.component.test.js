import React from 'react';
import { shallow } from 'enzyme';
import Datalist from './Datalist.component';

const schema = {
	autoFocus: true,
	description: 'This is my datalist',
	disabled: false,
	placeholder: 'Type here',
	readOnly: false,
	restricted: true,
	title: 'My List',
	titleMap: [
		{ name: 'Foo', value: 'foo' },
		{ name: 'Bar', value: 'bar' },
		{ name: 'Lol', value: 'lol' },
	],
	type: 'string',
	schema: {
		type: 'string',
	},
};

describe('Datalist component', () => {
	it('should render', () => {
		// when
		const wrapper = shallow(
			<Datalist
				id={'my-datalist'}
				isValid
				errorMessage={'This should be correct'}
				onChange={jest.fn()}
				onFinish={jest.fn()}
				onTrigger={jest.fn()}
				schema={schema}
				value={'foo'}
			/>,
		);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	describe('onChange', () => {
		it('should call props.onChange && props.onFinish', () => {
			// when
			const props = {
				onChange: jest.fn(),
				onFinish: jest.fn(),
				onTrigger: jest.fn(),
				schema: {
					type: 'string',
					schema: {
						type: 'string',
					},
				},
			};
			const wrapper = shallow(<Datalist {...props} />);
			const selectedValue = { label: 'Bar', value: 'bar' };
			const event = { type: 'change' };
			wrapper.instance().onChange(event, selectedValue);

			// then
			expect(props.onChange).toHaveBeenCalled();
			expect(props.onFinish).toHaveBeenCalled();
			expect(props.onFinish.mock.calls[0][0]).toEqual(event);
			expect(event.type).toBe('change');
			expect(props.onChange.mock.calls[0][1].value).toBe('bar');
		});
		it('should support undefined value', () => {
			// when
			const props = {
				onChange: jest.fn(),
				onFinish: jest.fn(),
				onTrigger: jest.fn(),
				schema: {
					type: 'string',
					schema: {
						type: 'string',
					},
				},
			};
			const event = { type: 'change' };
			const payload = undefined;
			const wrapper = shallow(<Datalist {...props} />);
			wrapper.instance().onChange(event, payload);
			// then
			expect(props.onChange).toHaveBeenCalledWith(event, { schema: props.schema });
		});
	});
	describe('onFocus', () => {
		it('should call onTrigger when triggers has onEvent="focus"', done => {
			// when
			const data = { titleMap: [{ name: 'Foo', value: 'foo' }] };
			const props = {
				onChange: jest.fn(),
				onFinish: jest.fn(),
				onTrigger: jest.fn(
					event =>
						new Promise(resolve => {
							// hack: to be sure we catch the setState after the promise
							setTimeout(() => {
								expect(event.target.state.isLoading).toBe(false);
								done();
							}, 0);
							return resolve(data);
						}),
				),
				schema: {
					type: 'string',
					schema: {
						type: 'string',
					},
					triggers: [
						{
							onEvent: 'focus',
						},
					],
				},
			};
			const wrapper = shallow(<Datalist {...props} />);
			wrapper.instance().onFocus({ type: 'focus' });

			// then
			expect(props.onTrigger).toHaveBeenCalled();
			expect(wrapper.state('isLoading')).toBe(true);
		});
	});

	describe('getTitleMap', () => {
		it('should return array from schema.titleMap', () => {
			// when
			const props = {
				onChange: jest.fn(),
				onFinish: jest.fn(),
				onTrigger: jest.fn(),
				schema,
			};
			const wrapper = shallow(<Datalist {...props} />);
			const options = wrapper.instance().getTitleMap();

			// then
			expect(options).toEqual([
				{ name: 'Foo', value: 'foo' },
				{ name: 'Bar', value: 'bar' },
				{ name: 'Lol', value: 'lol' },
			]);
		});
		it('should give priority to state.titleMap', () => {
			// when
			const props = {
				onChange: jest.fn(),
				onFinish: jest.fn(),
				onTrigger: jest.fn(),
				schema,
			};
			const wrapper = shallow(<Datalist {...props} />);
			wrapper.setState({
				titleMap: [{ name: 'Hello', value: 'hello' }, { name: 'World', value: 'world' }],
			});
			const options = wrapper.instance().getTitleMap();

			// then
			expect(options).toEqual([
				{ name: 'Hello', value: 'hello' },
				{ name: 'World', value: 'world' },
			]);
		});
		it('should add props.value to the titleMap', () => {
			// when
			const props = {
				onChange: jest.fn(),
				onFinish: jest.fn(),
				onTrigger: jest.fn(),
				schema,
				value: 'hello',
			};
			const wrapper = shallow(<Datalist {...props} />);
			const titleMap = wrapper.instance().getTitleMap();

			// then
			expect(titleMap).toEqual([
				{ name: 'Foo', value: 'foo' },
				{ name: 'Bar', value: 'bar' },
				{ name: 'Lol', value: 'lol' },
				{ name: 'hello', value: 'hello' },
			]);
		});
	});
});
