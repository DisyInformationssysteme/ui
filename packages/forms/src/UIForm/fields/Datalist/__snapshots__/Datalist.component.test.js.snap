// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datalist component should render 1`] = `
<FieldTemplate
  description="This is my datalist"
  errorMessage="This should be correct"
  id="my-datalist"
  isValid={true}
  label="My List"
  required={undefined}
>
  <Datalist component
    autoFocus={true}
    description="This is my datalist"
    disabled={false}
    id="my-datalist-select"
    input={true}
    multiSection={false}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type here"
    readOnly={false}
    restricted={true}
    titleMap={
      Array [
        Object {
          "name": "Foo",
          "value": "foo",
        },
        Object {
          "name": "Bar",
          "value": "bar",
        },
        Object {
          "name": "Lol",
          "value": "lol",
        },
        Object {
          "name": "foo",
          "value": "foo",
        },
      ]
    }
    value="foo"
  />
</FieldTemplate>
`;
